Mathlib/FieldTheory/Differential/Basic.lean

```diff
@@ -0,0 +1,146 @@
+as a preparation for Liouville's theorem.
+-/
+
+open Differential algebraMap
+
+variable {R : Type*} [Field R] [Differential R] (a b : R)
+
+/--
+The logarithmic derivative of a is a′ / a.
+-/
+def logd : R := a′ / a
+
+@[simp]
+lemma logd_zero : logd (0 : R) = 0 := by
+  unfold logd
+  simp
+
+@[simp]
+lemma logd_one : logd (1 : R) = 0 := by
+  unfold logd
+  simp
+
+lemma logd_add (h₁ : a ≠ 0) (h₂ : b ≠ 0) : logd a + logd b = logd (a * b) := by
+  unfold logd
+  field_simp
+  ring
+
+lemma logd_sub (h₁ : a ≠ 0) (h₂ : b ≠ 0) : logd a - logd b = logd (a / b) := by
+  unfold logd
+  field_simp [Derivation.leibniz_div]
+  ring
+
+lemma nat_mul_logd (a : ℕ) : a * logd b = logd (b ^ a) := by
+  by_cases h : b = 0
+  · cases a <;> simp [h]
+  induction a with
+  | zero => simp
+  | succ a h2 =>
+  rw [Nat.cast_add, Nat.cast_one, add_mul, one_mul, h2, logd_add, pow_succ] <;>
+  simp [h]
+
+lemma logd_eq_zero_iff : logd a = 0 ↔ a′ = 0 :=
+  ⟨fun h ↦ by simp only [logd, div_eq_zero_iff] at h; rcases h with h|h <;> simp [h],
+  fun h ↦ by unfold logd at *; simp [h]⟩
+
+lemma Multiset.logd_sum (ι : Type*) (s : Multiset ι) (f : ι → R) (h : ∀ x ∈ s, f x ≠ 0) :
+    (s.map (fun x ↦ logd (f x))).sum = logd (s.map f).prod := by
+  induction s using Multiset.induction_on
+  · simp
+  · rename_i h₂
+    simp only [Function.comp_apply, Multiset.map_cons, Multiset.sum_cons, Multiset.prod_cons]
+    rw [h₂]
+    apply logd_add
+    simp [h]
+    apply Multiset.prod_ne_zero
+    all_goals simp_all
+
+lemma logd_sum (ι : Type*) (s : Finset ι) (f : ι → R) (h : ∀ x ∈ s, f x ≠ 0) :
+    ∑ x ∈ s, logd (f x) = logd (∏ x ∈ s, f x) := Multiset.logd_sum _ _ _ h
+
+lemma logd_sum_zero (ι : Type*) (s : Finset ι) (f : ι → R) (h : ∀ x ∈ s, f x = 0) :
+    ∑ x ∈ s, logd (f x) = logd (∏ x ∈ s, f x) := by
+  unfold logd
+  simp_all
+
+lemma logd_algebraMap {F K : Type*} [Field F] [Field K] [Differential F] [Differential K]
+    [Algebra F K] [DifferentialAlgebra F K]
+    (a : F) : logd (algebraMap F K a) = algebraMap F K (logd a) := by
+  unfold logd
+  simp [deriv_algebraMap]
+
+@[norm_cast]
+lemma algebraMap.coe_logd {F K : Type*} [Field F] [Field K] [Differential F] [Differential K]
+    [Algebra F K] [DifferentialAlgebra F K]
+    (a : F) : logd a = logd (a : K) := (logd_algebraMap a).symm
+
+universe u v
+
+/--
+A field extension `K / F` is Liouville if, whenever an element a ∈ F can be written as
+`a = v + ∑ cᵢ * logd uᵢ` for `v, cᵢ, uᵢ ∈ K` and `cᵢ` constant, it can also be written in that
+way with `v, cᵢ, uᵢ ∈ F`.
+-/
+class IsLiouville (F : Type u) (K : Type*) [Field F] [Field K] [Differential F] [Differential K]
+    [Algebra F K] [DifferentialAlgebra F K] : Prop where
+  is_liouville (a : F) (ι : Type u) [Fintype ι] (c : ι → F) (hc : ∀ x, (c x)′ = 0)
+    (u : ι → K) (v : K) (h : a = ∑ x, c x * logd (u x) + v′) :
+    ∃ (ι' : Type u) (_ : Fintype ι') (c' : ι' → F) (_ : ∀ x, (c' x)′ = 0)
```

Comment:
```suggestion
    ∃ (ι' : Type) (_ : Fintype ι') (c' : ι' → F) (_ : ∀ x, (c' x)′ = 0)
```
since all fintypes are `0`-small. First thing you should do is show that this is equivalent to `∃ (ι' : Type u) (_ : Fintype ι') (c' : ι' → F) (_ : ∀ x, (c' x)′ = 0), ...` for any universe `u`. It will save you from the universe gymnastic you're currently performing in `IsLiouville.trans`

Reply:
Should I also have `(ι : Type)` as the argument?

Reply:
Yes, for the same reason